### ISSUE TITLE
Questions about discrepancies between the paper protocol and the released implementation

### ISSUE THREAD

**Issue by Credica** [TTomilin/MEAL#10]
Hello, and thank you for releasing MEAL. I am trying to reproduce the main Overcooked results from *MEAL: A Benchmark for Continual Multi-Agent Reinforcement Learning*. While comparing the paper with the released implementation, I found several protocol and metric differences that seem large enough to materially affect the results.

I inspected commit [`8a9fff5502c5318a2872db2d57d29fade97a1b71`](https://github.com/TTomilin/MEAL/commit/8a9fff5502c5318a2872db2d57d29fade97a1b71). Some of these may simply mean that the paper experiments used command-line overrides that are not recorded in the repository. Could you please clarify the intended reproduction configuration and, if possible, provide the exact commands/config files used for Tables 2–3 and Figure 16?

## Summary

| Item | Paper | Released implementation at the commit above |
|---|---|---|
| Evaluation interval | Section 5 says every 100 updates; Table 6 says every 5 updates | IPPO/HAPPO default to 5, while MAPPO defaults to 75 |
| Sequence length | Table 6: 20-task base sequence | IPPO/MAPPO/HAPPO default to 10; README examples also use shorter sequences |
| Reward-shaping horizon | Table 6: `2.5e6`, linearly annealed to zero | Config says `2.5e7`, but IPPO/MAPPO actually use `steps_per_task / 2` (`5e7` with the paper budget); HAPPO keeps the coefficient at 1.0 |
| Learning rate | Table 6: linear `1e-3 -> 1e-4` | `anneal_lr=False` by default; the optional schedule targets approximately zero rather than `1e-4` |
| Optimizer at task boundaries | Appendix C.2 says optimizer state is carried over | `reset_optimizer=True` by default and the state is reinitialized for every task |
| Evaluation episodes | Table 6: 10 | `eval_num_episodes=5` is defined but not used; IPPO/MAPPO use `seq_length`, and HAPPO uses `num_envs` |
| Shared rewards | Appendix B/N describe dense and sparse rewards as shared team rewards | Default dense mode shares delivery reward but keeps shaping agent-specific; sparse mode leaves delivery reward agent-specific |
| Continual-learning metrics | Equations (1)–(5) define normalized soup delivery, weighted forgetting, and FT with denominator `1 - AUC_baseline` | Result scripts use multiple different definitions; the main table script uses `AUC_baseline` as the FT denominator |

## 1. Evaluation interval is inconsistent within the paper

Section 5 states that the policy is evaluated **every 100 updates** using 10 episodes. However, Table 6 states an evaluation interval of **every 5 policy updates**, also using 10 episodes.

The released defaults are 5 for [IPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L130-L138) and [HAPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/happo.py#L185-L193), but 75 for [MAPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/mappo.py#L126-L134).

Which evaluation interval was used for the reported results: 5, 75, or 100 updates?

## 2. The paper uses a 20-task base sequence, while the released defaults use 10

Table 6 specifies `|T| = 20` for the base sequence, and the main figures/tables display 20 tasks. The released defaults set `seq_length=10` in [IPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L105-L113), [MAPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/mappo.py#L101-L109), and [HAPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/happo.py#L160-L168). The README quick-start command also explicitly uses `--seq-length 10`.

I assume the paper runs passed `--seq-length 20`. Is that correct for all results in Tables 2–3 and Figure 16?

## 3. The configured and effective reward-shaping horizons differ from Table 6

Table 6 reports a reward-shaping horizon of `2.5e6` environment steps, linearly annealed to zero.

In the released IPPO configuration, `reward_shaping_horizon` defaults to `2.5e7` ([config](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L56-L59)), but the field is not used by the training schedule. Instead, the implementation sets the effective horizon to `steps_per_task / 2` ([IPPO lines 390–395](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L390-L395); [MAPPO lines 413–416](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/mappo.py#L413-L416)). With `steps_per_task=1e8`, this is `5e7`, i.e. 20 times the Table 6 value.

HAPPO also uses `steps_per_task / 2`, but its Optax schedule has both `init_value=1.0` and `end_value=1.0`, so shaping never anneals ([HAPPO lines 547–552](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/happo.py#L547-L552)).

What was the effective horizon in the paper experiments: `2.5e6`, `2.5e7`, or `5e7`? Was HAPPO intended to use the same annealing schedule as IPPO/MAPPO?

## 4. The released default does not use the learning-rate schedule from Table 6

Table 6 specifies linear LR annealing from `1e-3` to `1e-4`. In all three trainers, `anneal_lr` defaults to `False` (for example, [IPPO lines 41–44](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L41-L44)), resulting in a constant `1e-3` learning rate.

When enabled, the IPPO schedule is `lr * (1 - progress)` and therefore targets approximately zero, not `1e-4` ([IPPO lines 305–311](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L305-L311)).

Were the reported experiments launched with `--anneal-lr true`? If so, was the intended endpoint zero or `1e-4`, and was annealing restarted for each task or applied over the complete sequence?

## 5. Optimizer state is reset despite Appendix C.2 saying it is carried over

Appendix C.2 states: “At task boundaries, the optimizer state and policy parameters are carried over.” However, `reset_optimizer=True` by default in [IPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L73-L83), [MAPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/mappo.py#L70-L81), and [HAPPO](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/happo.py#L128-L138).

The optimizer state is then reinitialized at each task boundary, e.g. [IPPO lines 380–383](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L380-L383) and [HAPPO lines 532–541](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/happo.py#L532-L541).

Should paper reproduction use `--reset-optimizer false`?

## 6. `eval_num_episodes` does not control the number of evaluation episodes

Table 6 specifies 10 evaluation episodes. The trainer configs define `eval_num_episodes=5`, but this field is not passed to `make_eval_fn`.

- IPPO passes `seq_length` as the number of parallel evaluation environments ([lines 365–366](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L365-L366)).
- MAPPO does the same ([lines 391–392](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/mappo.py#L391-L392)).
- HAPPO passes the training `num_envs`, whose default is 2048 ([lines 515–518](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/happo.py#L515-L518)).

`make_eval_fn` uses this argument to split reset keys and average parallel rollouts ([evaluation.py lines 20–44](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/evaluation.py#L20-L44)). Thus a 20-task IPPO/MAPPO run evaluates 20 episodes per task, while default HAPPO evaluates 2048.

Was the intended value 10 for every algorithm?

## 7. “Dense Shared” and “Sparse Shared” do not appear fully shared in the trainer

Appendix B describes the reward as a shared team reward containing delivery and shaping terms. Appendix N calls the compared settings “Dense Shared”, “Dense Individual”, and “Sparse Shared”.

The environment emits per-agent delivery and shaping values ([overcooked.py lines 484–492](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/meal/env/overcooked.py#L484-L492)). In the default trainer path, delivery rewards are summed and shared, but each agent receives only its own shaped reward ([IPPO lines 457–465](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L457-L465)). In sparse mode, the trainer performs no transformation, leaving delivery rewards assigned only to the agent that delivered the soup ([IPPO lines 446–450](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/ippo.py#L446-L450)).

Could you clarify the intended definitions?

1. Dense Shared = shared delivery + **sum of all agents' shaping** for every agent, or shared delivery + **each agent's own shaping**?
2. Sparse Shared = summed delivery reward for every agent, or individual delivery reward?

### Forward Transfer

Equation (5) defines each term as:

```text
(AUC_i - AUC_i^baseline) / (1 - AUC_i^baseline)
```

The main numerical table script instead computes:

```python
(auc_cl - auc_baseline) / max(abs(auc_baseline), epsilon)
```

See [`results_table.py` lines 327–331](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/results/numerical/results_table.py#L327-L331). Other scripts also divide by `auc_baseline`, for example [`reward_settings.py` lines 201–206](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/results/numerical/reward_settings.py#L201-L206).

Which FT formula generated Table 2 and the FT values elsewhere in the paper?

### Forgetting

Equation (3) defines weighted, normalized post-training forgetting relative to performance at the task boundary. `results_table.py` implements a close version of that definition ([lines 42–108](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/results/numerical/results_table.py#L42-L108)). However, other result scripts use an unnormalized peak-to-final drop, e.g. [`algorithm_comparison.py` lines 79–87](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/results/numerical/algorithm_comparison.py#L79-L87) and [`reward_settings.py` lines 210–217](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/results/numerical/reward_settings.py#L210-L217).

Were all reported forgetting values calculated with Equation (3), or did different tables use different definitions?

### Aggregate normalized soup metric

`add_eval_metrics` correctly logs each task's scaled soup value as `avg_soups[i] / max_soup_dict[i]`, but `Evaluation/Soup_Scaled_Over_All_Env` sums the **unscaled** `avg_soups` values ([utils.py lines 97–104](https://github.com/TTomilin/MEAL/blob/8a9fff5502c5318a2872db2d57d29fade97a1b71/experiments/utils.py#L97-L104)). It is also a sum rather than the task mean in Equation (1).

Was this aggregate key used for reported Average Soup Delivery, or were the per-task scaled values normalized and averaged later by the result scripts?

## Requested reproduction details

To make the published numbers reproducible, it would be very helpful to have:

1. The exact IPPO/MAPPO/HAPPO commands or serialized configs used for Tables 2–3 and Figure 16.
2. Confirmation of the effective reward-shaping horizon and shared-reward definitions.
3. Confirmation of whether optimizer state was retained and how LR annealing was applied.
4. Confirmation of the evaluation interval and number of episodes for each algorithm.
5. The exact analysis script/formulas used for `S`, `F`, and `FT` in each reported table.

Thank you for your help and for making the benchmark available.

**Comment by TTomilin** [comment]
Hi, thank you for the extremely thorough review. This caught several bugs, and I indeed found a couple of places where the text in the paper is ambiguous or doesn't exactly match what we ran. I'll go through each point below. Fixes are on the `paper-reproduction-issues` branch.

## 1. Evaluation interval

Every 5 updates is correct, for all the algorithms. Table 6 is right, the "every 100 updates" in Section 5 was not updated. We reduced the interval when we increased the `num_envs` for our final experiments.

## 2. Sequence length

You assumed correctly. All Table 2–3 / Figure 16 results used `--seq-length 20`. The `seq_length=10` defaults in the repo (and the README quick-start example) were just meant as examples. Default is now 20 everywhere to avoid this confusion.

## 3. Reward-shaping horizon

The horizon used for the reported results was ultimately `steps_per_task / 2` for every algorithm (`5e7` with `steps_per_task=1e8`), **not** the `2.5e6` in Table 6. We increased the `steps_per_task` for our experiments from 1e7 --> 1e8, but didn't update Table 6 to reflect this.

Now if you pass `--reward-shaping-horizon`, it's used; if you don't (the default, `None`), it falls back to `steps_per_task / 2`, matching our results. HAPPO now also follows this by default.

## 4. Learning rate

We used `--anneal-lr true`, annealing `1e-3 → 1e-4`, matching what's in Table 6. We now unified this across algos with `optax.linear_schedule(cfg.lr, cfg.lr_end, total_grad_steps)` with `lr_end` defaulting to `1e-4`.

## 5. Optimizer state at task boundaries

This one's the other way around: `reset_optimizer=True` (the default) is what we actually used for every experiment. The optimizer state (Adam's first/second moment estimates) is indeed reset at each task boundary, only the network parameters carry over. Appendix C.2's description ("the optimizer state and policy parameters are carried over") is thus not fully accurate, as only the parameters are carried over. This is also the more common choice for continual learning. Carrying over stale Adam moment estimates from a very different task's gradient distribution tends to hurt early adaptation to the new task. No code change here.

## 6. Evaluation episodes

`eval_num_episodes` was defined as `10` (default, same as in Table 6), but wasn't passed into `make_eval_fn` for any algorithm. It is now fixed and usable. Importantly, though, this had no effect on any number in the paper. Table 2/3 and Figure 16 were computed entirely from the training-time logged curves (the periodic `Evaluation/Soup_Scaled/{task}_{layout}` series and the `training_soup.json`/`Soup/scaled` series), not from a separate post-hoc evaluation run. The result scripts never touch `eval_num_episodes`'s effective value at all.

## 7. Shared vs. individual rewards

Your reading of the code was right, and this was a bug. We fixed the definitions, matching the paper's intent (Appendix B/N):

- **Dense Shared**: delivery reward summed across agents and broadcast to all, **and** shaping reward also summed across agents and broadcast to all (previously: shaping stayed per-agent, only delivery was shared).
- **Dense Individual**: unchanged, both delivery and shaping stay per-agent.
- **Sparse Shared**: delivery reward summed across agents and broadcast to all (previously: delivery reward was left per-agent even in the "shared" sparse setting, i.e. sparse mode wasn't actually shared at all).

## 8. Metrics

### Forward Transfer

The `1 - AUC_baseline` in the denominator of Equation 5 doesn't work for MEAL. That denominator assumes scores are bounded in `[0, 1]`, but MEAL's soup-delivery score can exceed 1.0, so `1 - AUC_baseline` can go negative or near-zero and blow up the ratio. The code's `AUC_baseline` denominator (not `1 - AUC_baseline`) is what was actually used for every reported FT value. 

### Forgetting

Here we mixed both formulas.
- The baseline results in **Table 2** (`results_table.py`) and Figure 4 (`forward_transfer_vs_forgetting.py`) use the weighted, normalized formula (Equation 3), a decayed-and-normalized post-training drop relative to end-of-training performance, clipped to `[0, 1]`.
- **Table 3** (`agents_comparison.py`), **Figure 16** (`algorithm_comparison.py`), and every other numerical/plotting script that reports Forgetting (`reward_settings.py`, `curriculum_comparison.py`, `observable_comparison.py`, `restrictions.py`, `partner_adaption.py`, `partner_adaption_arch_table.py`) instead the unnormalized peak-minus-final-window drop.

Both formulas now live in `experiments/results/plotting/utils/metrics.py` (`calculate_curve_based_forgetting` for the weighted/normalized one, `calculate_peak_final_forgetting` for the peak-minus-final one), and every script that reports Forgetting now takes a `--forgetting-formula {weighted,peak_final}` CLI flag, defaulting to whichever formula actually produced that script's numbers.

### Aggregate normalized soup metric

Indeed, `Evaluation/Soup_Scaled_Over_All_Env` is exactly the bug you described. It sums the unscaled `avg_soups` values (identical to `Evaluation/Soup_Over_All_Env` one line above it. However, this was just a convenience metric with no effect on anything reported. The reported "Average Soup Delivery" numbers are computed independently by the result scripts from the per-task `Evaluation/Soup_Scaled/{task}_{layout}` series (downloaded separately as `{i}_soup.json`), taking the mean across tasks of the final-eval value, per seed, aggregated with a CI.

Thanks again for digging into this in such detail! Happy to answer any follow-ups if anything above needs more specifics or if you encounter any other issues.
